internal/node/labeler/labeler.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package labeler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	hlaiv1alpha1 "github.com/HabanaAI/habana-ai-operator/api/v1alpha1"
+	"github.com/HabanaAI/habana-ai-operator/internal/constants"
+	s "github.com/HabanaAI/habana-ai-operator/internal/settings"
+)
+
+const (
+	nodeLabelerServiceAccount = "node-labeler"
+	nodeLabelerSuffix         = "node-labeler"
+	nodeLabelerLimitsCpu      = "1"
+	nodeLabelerLimitsMemory   = "200Mi"
+	nodeLabelerRequestsCpu    = "100m"
+	nodeLabelerRequestsMemory = "200Mi"
+)
+
+//go:generate mockgen -source=labeler.go -package=labeler -destination=mock_labeler.go
+
+type Reconciler interface {
+	ReconcileNodeLabeler(ctx context.Context, dc *hlaiv1alpha1.DeviceConfig) error
+	DeleteNodeLabeler(ctx context.Context, dc *hlaiv1alpha1.DeviceConfig) error
+	ReconcileNodeLabelerDaemonSet(ctx context.Context, dc *hlaiv1alpha1.DeviceConfig) error
+	SetDesiredNodeLabelerDaemonSet(ds *appsv1.DaemonSet, cr *hlaiv1alpha1.DeviceConfig) error
+	DeleteNodeLabelerDaemonSet(ctx context.Context, dc *hlaiv1alpha1.DeviceConfig) error
+}
+
+type NodeLabelerReconciler struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func NewReconciler(c client.Client, s *runtime.Scheme) *NodeLabelerReconciler {
+	return &NodeLabelerReconciler{
+		client: c,
+		scheme: s,
+	}
+}
+
+func GetNodeLabelerName(cr *hlaiv1alpha1.DeviceConfig) string {
```

Comment:
We could make this function private, as it's only used internally in this package:

```go
func getNodeLabelerName(cr *hlaiv1alpha1.DeviceConfig) string
```